src_cpp/opennnBridge/train.h

```diff
@@ -0,0 +1,210 @@
+#pragma once 
+
+//#include <iostream>
+#include <vector>
+#include <string>
+#include "ModelParams.h"
+#include <map>
+//#include <eigen3/Eigen/Core>
+#include "../opennn/opennn/opennn.h"
+
+#include "definitionsNN.h"
+
+using namespace OpenNN;             
+
+struct TrainNN {
+
+    long int mid;
+    int optimization_method;
+    int lose_method;
+    Eigen::Tensor<float,2> data;
+
+    ErlNifTid tid;
+    ErlNifPid pid;
+};
+ 
+static void* trainFun(void* arg){ 
+         printf("start train\n");
+         TrainNN* TrainNNptr = (TrainNN*)arg;
+         int loss_val;
+         ErlNifEnv *env = enif_alloc_env();
+         printf("data_set \n");
+         //cout << TrainNNptr->data <<std::endl;
+         DataSet data_set;
+         data_set.set_data(TrainNNptr->data);
+         printf("end data_set \n");
+         //MeanSquaredError mean_squared_error;
+         
+         
+         // Get the singleton instance
+         opennnBridgeController *s = s->GetInstance();
+         printf("Get the model from the singleton \n");
+         // Get the model from the singleton
+         cout << "model ID is " <<std::endl;
+         cout << TrainNNptr->mid << std::endl;
+         std::shared_ptr<OpenNN::NeuralNetwork> neural_network = s-> getModelPtr(TrainNNptr->mid);
+         //NeuralNetwork neural_network = *(s-> getModelPtr(mid));
+
+
+         printf("end Get the model from the singleton \n");
+         
+
+         //cout << neural_network->get_layers_number() <<std::endl;
+         //TrainingStrategy training_strategy(&neural_network ,&data_set);
+
+         // ask david
+         TrainingStrategy training_strategy(&(*(s-> getModelPtr(TrainNNptr->mid))) ,&data_set);
```

Comment:
Ask me

Reply:
training_strategy don't love share pointer , so i did it : &(*(share pointer)) , is it ok?

Reply:
You suppose to use sharePtr->get() which returns the raw pointer